### PR TITLE
Support async  using promises

### LINKED ISSUE #157: support async localGeocoder

### PR DESCRIPTION
**Pull request by yuletide**
This PR adds support for asynchronous `localgeocoder` functions using Promises. 

As it stands, this will likely require additional polyfills to prevent breakage in older browsers. Not sure the best approach for that but open to ideas before we merge.

closes #157

### CODE REVIEW

**Comment by andrewharvey** [comment]
In hind sight calling this localGeocoder was probably a mistake on my part, as it's now gone past just local sources into any kind of 3rd party supplement data source, local or remote. Not sure if it's best to leave that be or to make a breaking change to address it.

Personally I'm not concerned about breaking API changes to make thing simpler and clearer.

**Comment by andrewharvey** [comment]
> As it stands, this will likely require additional polyfills to prevent breakage in older browsers. Not sure the best approach for that but open to ideas before we merge.

I'm okay with that, so long as it's marked a breaking change.

https://github.com/mapbox/mapbox-sdk-js#installation does a good job of providing instructions for how to continue to support older browsers by using a polyfill, I think we should also provide those instructions.

File: lib/index.js [hunk 1]
```diff
@@ -155,20 +155,26 @@ MapboxGeocoder.prototype = {
...
     this._eventEmitter.emit('loading', { query: searchInput });
-    var request = this.mapboxClient.geocodeForward(searchInput, this.options);
-
-    var localGeocoderRes = [];
+    var mbGeocoderRes = this.mapboxClient.geocodeForward(searchInput, this.options);
+    var request;
+    
+    var localGeocoderRes = Promise.resolve([]); // pass in empty array promise for later handling.
     if (this.options.localGeocoder) {
+      // console.log("we have a local geocoder", this.options.localGeocoder);
```
Comment (andrewharvey) [inline lib/index.js]: I'd remove this from the final version.

File: lib/index.js [hunk 2]
```diff
@@ -155,20 +155,26 @@ MapboxGeocoder.prototype = {
...
       localGeocoderRes = this.options.localGeocoder(searchInput);
-      if (!localGeocoderRes) {
-        localGeocoderRes = [];
-      }
     }
+    request = Promise.all([localGeocoderRes, mbGeocoderRes]);
+
+    request.then(function (responses) {
+      // console.log(responses);
```
Comment (andrewharvey) [inline lib/index.js]: I'd remove this from the final version.

**Comment by andrewharvey** [comment]
Thanks @yuletide this is a useful feature, looking forward to it.

I don't think circleci run's the unit tests, is that right? Let's make sure they are all still passing.

**Comment by yuletide** [comment]
Working on this again. 

> In hind sight calling this localGeocoder was probably a mistake on my part, as it's now gone past just local sources into any kind of 3rd party supplement data source, local or remote. Not sure if it's best to leave that be or to make a breaking change to address it.

@andrewharvey per your comment earlier I was thinking we could leave `localgeocoder` as is, and create a new option `externalGeocoder` that takes a Promise and has a slightly more intuitive name.

**Comment by andrewharvey** [comment]
> @andrewharvey per your comment earlier I was thinking we could leave localgeocoder as is, and create a new option externalGeocoder that takes a Promise and has a slightly more intuitive name.

That sounds reasonable to me.

**Comment by styledev** [comment]
@andrewharvey Where do we stand on getting this code over the finish line? It is much needed.

**Comment by andrewharvey** [comment]
hi @styledev I'm not sure if @yuletide is still interested in working on this, if not happy for you to take over with another PR building on this if you want to contribute.

- [ ] rebase against master
- [ ] change this to `externalGeocoder`

Then I can do another review.

**Comment by yuletide** [comment]
@styledev Glad there is renewed interest in this! If you are looking to contribute please dive in. I've been pulled into some other project and won't have time for this for the next few months at least but happy to help with questions and PR's

**Comment by styledev** [comment]
@yuletide / @andrewharvey I worked out an alternative solution for now to get past launch but would like to dive in and see if I can get it over the finish line. Aside from the failed checks and the two items Andrew noted is there anything else that needs to be done?

**Comment by andrewharvey** [comment]
> Aside from the failed checks and the two items Andrew noted is there anything else that needs to be done?

I'm not sure, but if you do work on a PR, I can review and that would identify anything else if there is.

**Comment by wsoeltz** [comment]
Are there any updates as to when this feature will be added? It would be very useful for extending Mapbox's capabilities

**Comment by andrewharvey** [comment]
> Are there any updates as to when this feature will be added? It would be very useful for extending Mapbox's capabilities

@wsoeltz I'm not sure if this is something @yuletide is still working on or timeframe, but given there hasn't been any recent work on this, I'd say feel free to pick this up and if you want to submit a new PR I'd be happy to help review.

**Comment by duebbert** [comment]
For those interested, I've implemented the same changes with the current master version. Some small fixes but it works for me(tm). Haven't had time yet to solve any potential issues and didn't look too closely at the code yet but it's in this commit for others to use: https://github.com/duebbert/mapbox-gl-geocoder/commit/ff66b647e1e1f77979c65f33067c7062c2950b4f

**Comment by andrewharvey** [comment]
an alternative proposal is at #385

**Comment by mpothier** [comment]
Closing this issue, as this functionality is encapsulated in the `externalGeocoder` introduced in #385

**Pull request closed without merging**